### ISSUE TITLE
Different accuracy results

### ISSUE THREAD

**Issue by sarahmahwash** [Shaddyjr/bee-image-classifier#2]
Hi. So I followed your code mostly but I got different accuracy scores for each model. My first two models had the same score and actually outperformed the Brightness model. What could be the reason?

And how can I feed a single image into this trained CNN to predict if it has varroa or not?

**Comment by Shaddyjr** [comment]
Hi @sarahm09. Hmm, well, there are so many factors that could be at play during the model training. If you're altering the network in any way, using a different number of epochs or modifying the data in any way that could result in different model performances. You mentioned you followed this code "mostly" - what was different? Also, what were the model accuracy scores? Were they off by a lot or just minor differences?

Regarding feeding in a new image and having the model make a prediction, I set up a (now deprecated) website. A user would be able to submit a novel image, the backend would transform the image to ensure the model is fed the right image dimensions, the model would make a prediction, and then the result would be returned to the user. See the code in [this repo](https://github.com/Shaddyjr/bee_classifier_website). This whole website setup is probably overboard for what you're looking for, though. You could just run some code to take an image, resize it to fit the input dimensions, and have your model make the prediction.

I'll look into getting the prediction website back up and running. This is an old project I'm no longer maintaining, but I was expecting the website to still be working.

**Comment by sarahmahwash** [comment]
Hey! Thanks for replying.
Well the test scores for the models were 99.5, 99.5 and 99.1. I didn't make any changes in the code honestly, I just wanted to see how something like this would work. I am very new to Python and ML. 

So if my first two models have the same test scores, how can I pick the more efficient one? What should I be looking for? And how can I justify that my NoTransform model has performed the best? I mean it had the least misclassifications and the best train and test scores but does it not make more sense that since brightness is adjusted the Bright model should have worked better?

How did you decide to use 50 Epochs and not 10 or 20?

**Comment by Shaddyjr** [comment]
I've updated this repo with the latest library versions, which does end up altering the model performance. I also found the "No Transformation" model to be best.

Regarding picking the "best" model - I'd go with a deeper analysis into which model seems to be making the most sense of the data. What errors are the models making? Do the convolutional layers seems to be focusing in on the parts of the image that make the most sense?

I went with 50 epochs as a nice round number and I don't mind torturing my computer 😅. Certainly after ~20 or so epochs we see the models don't improve much, so reducing the # of epochs or using some early stopping would be practical.

**Comment by sarahmahwash** [comment]
Awesome! Thanks so much for clarifying. I also am curious as to why you decided to normalize your images between -0.5 and 0.5. Most sources recommend 0,1 or -1,1. Does it make much of a difference?

**Comment by Shaddyjr** [comment]
That's a great question!
It's been so long since I recall that particular decision I honestly couldn't say why I chose to normalize that way. (Take this with a grain of salt) If I recall correctly, I was trying everything to get these models to perform well and using a -.5-.5 normalization seemed to help. I believe I also either read about the normalization via research or my instructor at the time recommended it.

This Bee Image Classifier was actually my first project creating a CNN. In an effort to optimize, I may have made a mistake with the normalization. After some cursory research, what you say is correct - using a 0-1 range is fairly standard and should probably apply to this project!

---
Also, now that I'm looking more at the project, I'm seeing a LOT that can be improved. Definitely try out the 0-1 normalization and see how it works for you. _I will note for this particular project, many of the images are of VERY poor quality, so trying to squeeze out more performance may just not be worthwhile_. After seeing that 99+% testing accuracy, I just stopped trying to improve the model.

**Comment by sarahmahwash** [comment]
I tried with 0,1 normalization, the models didn't show much improvement so I'm gonna stick with your parameters lol. 

Yeah the accuracy here is pretty good but I'm wondering how to get accurate results when feeding individual unknown images into the models. I took random images of infected bees from different journal papers and tried predicting with all the models. Right now, they're all doing rather poorly. What can I try changing up? 

(Really sorry for so many questions btw, I don't have many sources to consult in real life lol)

**Comment by Shaddyjr** [comment]
No worries on the questions! This stuff is complicated.

To answer your question "What can I try changing up?" - it would ultimately be to train your model on A LOT of high quality images like the ones that would be used in practice.

---
### I'd say there are 3 important takeaways from this project:
1. This project was meant to serve as a proof of concept (POC), and wasn't intended to be used in the field with actual real images.
    - Real images would most likely be of MUCH higher quality than the training data, both in dimension and in subject matter. High dimension images fed to the model would have to be significantly reduced in quality and would certainly get distorted making it unlikely for the model to perform well.
    - Real images would likely include multiple bees, which the model wouldn't be able to handle. 
    - As such, __this project was just to explore the idea to see if it would be possible to do bee image classification - and I'd say the model does show promise!__
2. __"Garbage in, Garbage out"__
    - The training data is VERY lacking. The dataset images are inconsistent, with many being distorted while others don't appear to show anything resembling a bee at all. 
    - Relatively speaking, the number of images is quite small for a CNN (though it's a practical amount for a POC). 
    - __So, I'm not surprised the model isn't performing well on the real world images you've tried - it wasn't trained on such images to begin with.__
3. I failed to understand the problem
    - I don't mention this in the project, since it showcases the POC well, but after completing this project I showed it to a bee keeping expert to get their thoughts.
    - They informed me that __if you can visually see mites within your colony, then you're already too late.__ Bee keepers have a number of strategies to combat varroa and preventing an infestation is MUCH easier than treating one.
    - __If I had done more research and better understood the problem I wouldn't have wasted so much time and effort on a problem that was already solved.__

**Comment by Shaddyjr** [comment]
As an aside, I got the web app back up and running!
1. Repo - https://github.com/Shaddyjr/bee_classifier_website
2. App - https://bee-image-classifier.herokuapp.com/

Will close this issue out in ~7 days if there are no more questions.

**Comment by Shaddyjr** [comment]
I've also gone ahead and updated the models with `BatchNormalization` and improved model stability during training. The "No Transformation" model still wins out, but only just.

I also fixed a bug where the best model was NOT getting saved and properly used throughout the notebook - the last model was getting used instead, which was often not the "best" found during the CNN grid search.
